lib/RateLimiterDynamo.js

```diff
@@ -0,0 +1,422 @@
+      if(data.Item) {
+        return new DynamoItem(
+          data.Item.key.S,
+          Number(data.Item.points.N),
+          Number(data.Item.expire.N)
+        );
+      } else {
+        return null;
+      }
+    }
+
+    /**
+     * Deletes an item from the table based on the given rlKey.
+     *
+     * @param {string} rlKey - The rlKey of the item to delete.
+     * @throws {Error} Throws an error if the table is not created yet.
+     * @return {boolean} Returns true if the item was successfully deleted, otherwise false.
+     */
+    async _delete(rlKey) {
+
+      if (!this.tableCreated) {
+        throw new Error('Table is not created yet');
+      }
+
+      const params = {
+        TableName: this.tableName,
+        Key: {
+          key: {S: rlKey}
+        },
+        ConditionExpression: 'attribute_exists(#k)',
+        ExpressionAttributeNames: {
+          '#k': 'key'  
+        }
+      }
+      
+      try {
+        const data = await this._client.deleteItem(params);
+        return data.$metadata.httpStatusCode === 200;
+      } catch(err) {
+        // ConditionalCheckFailed, item does not exist in table
+        if (err.__type.includes('ConditionalCheckFailedException')) {
+          return false;
+        } else {
+          throw err;
+        }
+      }
+
+    }
+
+    /**
+     * Implemented with DynamoDB Atomic Counters. 3 calls are made to DynamoDB but each call is atomic.
+     * From the documentation: "UpdateItem calls are naturally serialized within DynamoDB,
+     * so there are no race condition concerns with making multiple simultaneous calls."
+     * See: https://aws.amazon.com/it/blogs/database/implement-resource-counters-with-amazon-dynamodb/
+     * @param {*} rlKey 
+     * @param {*} points 
+     * @param {*} msDuration 
+     * @param {*} forceExpire 
+     * @param {*} options 
+     * @returns
+     */
+    async _upsert(rlKey, points, msDuration, forceExpire = false, options = {}) {
+
+      if (!this.tableCreated) {
+        throw new Error('Table is not created yet');
+      }
+
+      const dateNow = Date.now();
+      /* -1 means never expire, DynamoDb do not support null values in number fields.
+         DynamoDb TTL use unix timestamp in seconds.
+      */
+      const newExpireSec = msDuration > 0 ? (dateNow + msDuration) / 1000 : -1;
+
+      // Force expire, overwrite points. Create a new entry if not exists
+      if (forceExpire) {
+        return await this._baseUpsert({
+          TableName: this.tableName,
+          Key: { key: {S: rlKey} },
+          UpdateExpression: 'SET points = :points, expire = :expire',
+          ExpressionAttributeValues: {
+            ':points': {N: points.toString()},
+            ':expire': {N: newExpireSec.toString()}
+          },
+          ReturnValues: 'ALL_NEW'
+        });
+      }
+
+      // First try update, success if entry does not exists
+      try {
+        
+        return await this._baseUpsert({
```

Comment:
@Daniel-97 Let's think about the first request. On working server with some load, it would fail in like 95% of requests. Of course, it depends on `duration` setting. Still, do you think, the update of the existing key should be called the first?

Also, the first and the third request look quite similar except `ConditionExpression`. Could we join both conditions in one? Like `expire <= :where_expire OR attribute_not_exists(points)`

Reply:
You are right, i have collapsed the first and the second update in one single update with the OR condition. I agree with you it is a better way to do it